parquet-hadoop/src/main/java/org/apache/parquet/filter2/dictionarylevel/DictionaryFilter.java

```diff
@@ -363,14 +363,54 @@ public Boolean visit(Not not) {
         "This predicate contains a not! Did you forget to run this predicate through LogicalInverseRewriter? " + not);
   }
 
+  private <T extends Comparable<T>, U extends UserDefinedPredicate<T>> Boolean visit(UserDefined<T, U> ud, boolean inverted) {
+    Column<T> filterColumn = ud.getColumn();
+    ColumnChunkMetaData meta = getColumnChunk(filterColumn.getColumnPath());
+
+    // Cannot make assumptions around UDPs and acceptance of null values
+    if (meta == null || hasNonDictionaryPages(meta)) {
+      return BLOCK_MIGHT_MATCH;
+    }
+
+    try {
+      Set<T> dictSet = expandDictionary(meta);
+      if (dictSet == null) {
+        return BLOCK_MIGHT_MATCH;
+      }
+
+      boolean allMatch = true;
+      for(T entry : dictSet) {
+        if (ud.getUserDefinedPredicate().keep(entry)) {
+          if (!inverted) {
+            return BLOCK_MIGHT_MATCH;
+          }
+        } else {
+          allMatch = false;
```

Comment:
I think this can have similar short-circuit logic. If the entry does not match but the predicate is inverted, then it does match and so does the entire block.

That would allow you to simplify the logic below, which cannot get to the third (else) case without hitting the short-circuit condition above.